der/src/tag.rs

```diff
@@ -271,85 +287,159 @@ impl Tag {
+            0x12 => Tag::NumericString,
+            0x13 => Tag::PrintableString,
+            0x14 => Tag::TeletexString,
+            0x15 => Tag::VideotexString,
+            0x16 => Tag::Ia5String,
+            0x17 => Tag::UtcTime,
+            0x18 => Tag::GeneralizedTime,
+            0x1A => Tag::VisibleString,
+            0x1B => Tag::GeneralString,
+            0x1E => Tag::BmpString,
+            0x30 => Tag::Sequence, // constructed
+            0x31 => Tag::Set,      // constructed
+            0x40..=0x7F => {
+                let (constructed, number) = parse_parts(first_byte, reader)?;
+
+                Tag::Application {
+                    constructed,
+                    number,
+                }
+            }
+            0x80..=0xBF => {
+                let (constructed, number) = parse_parts(first_byte, reader)?;
+
+                Tag::ContextSpecific {
+                    constructed,
+                    number,
+                }
+            }
+            0xC0..=0xFF => {
+                let (constructed, number) = parse_parts(first_byte, reader)?;
+
+                Tag::Private {
+                    constructed,
+                    number,
+                }
+            }
+            // universal tag in long form
+            0x1F => return Err(ErrorKind::TagNumberInvalid.into()),
+            byte => return Err(ErrorKind::TagUnknown { byte }.into()),
+        };
+
+        Ok(tag)
     }
 }
 
-impl From<Tag> for u8 {
-    fn from(tag: Tag) -> u8 {
-        tag.octet()
+fn ensure_multi_byte(number: u32) -> Result<TagNumber> {
+    if number >= u32::from(TagNumber::MASK) {
+        Ok(TagNumber::new(number))
+    } else {
+        // must be in short form
+        Err(ErrorKind::TagNumberInvalid.into())
     }
 }
 
-impl From<&Tag> for u8 {
-    fn from(tag: &Tag) -> u8 {
-        u8::from(*tag)
+fn parse_parts<'a, R: Reader<'a>>(first_byte: u8, reader: &mut R) -> Result<(bool, TagNumber)> {
+    let constructed = first_byte & CONSTRUCTED_FLAG != 0;
+    let first_number_part = first_byte & TagNumber::MASK;
+
+    if first_number_part != TagNumber::MASK {
+        return Ok((constructed, TagNumber::new(first_number_part.into())));
     }
-}
 
-impl<'a> Decode<'a> for Tag {
-    type Error = Error;
+    let mut multi_byte_tag_number = 0;
 
-    fn decode<R: Reader<'a>>(reader: &mut R) -> Result<Self> {
-        reader.read_byte().and_then(Self::try_from)
+    for _ in 0..Tag::MAX_SIZE - 2 {
+        multi_byte_tag_number <<= 7;
+
+        let byte = reader.read_byte()?;
+        multi_byte_tag_number |= u32::from(byte & 0x7F);
+
+        if byte & 0x80 == 0 {
+            return Ok((constructed, ensure_multi_byte(multi_byte_tag_number)?));
+        } else if byte == 0x80 {
+            // leading zeros in long form
+            return Err(ErrorKind::TagNumberInvalid.into());
+        }
+    }
+
+    let byte = reader.read_byte()?;
+    if multi_byte_tag_number > u32::MAX >> 7 || byte & 0x80 != 0 {
```

Comment:
This overflow check doesn't work, since we've already shifted too far before we get to this point. Checking for leading zeros seems simpler (see suggestion below).

There should be an overflow test case.